zerver/lib/markdown/__init__.py

```diff
@@ -1595,9 +1595,59 @@ def __init__(self, compiled_re: Pattern, md: markdown.Markdown) -> None:
 
 
 class AutoLink(CompiledPattern):
+    # NOTE: This value may vary depending on the size of repository, but we don't
+    # have that information and thus we pick a value that's large enough.
+    COMMIT_ID_PREFIX_LENGTH = 10
+
+    def shorten_links(self, href: str) -> Optional[str]:
+        parts = urllib.parse.urlparse(href)
+        scheme, netloc, path, params, query, fragment = parts
+        if scheme == "https" and netloc in ["github.com"]:
+            # Split the path to extract our 4 variables.
+
+            # To do it cleanly without multiple if branches based on which of these
+            # variables are present, we here add a list of ["", "", ""...]
+            # to the result of path.split, which at worst can be []. We also remove
+            # the first empty string we'd get from "/foo/bar".split("/").
+
+            # Example path: "/foo/bar" output: ["foo", "bar", "", "", ""]
+            #         path: ""         output: ["", "", "", "", ""]
+            organisation, repository, artifact, value, remaining_path = (
+                path.split("/", 5)[1:] + [""] * 5
+            )[:5]
+
+            # Decide what type of links to shorten.
+            if not organisation or not repository or not artifact or not value:
+                return None
+            repo_short_text = "{}/{}".format(organisation, repository)
+            # Some urls like https://github.com/zulip/zulip-mobile/pull/13168/commits can
+            # have some components after the artifact value as well. We should append that
+            # to the result if it is present.
+            remaining_text = ""
+            if remaining_path:
+                remaining_text += "/" + remaining_path
+            if fragment:
+                remaining_text += "#" + fragment
```

Comment:
I'm not sure it makes sense to shorten links of these forms -- I don't think GitHub itself does so either (correct me if I'm wrong for any of these examples; it should be easy enough to edit a GitHub comment to include all your test cases below to verify).  I think we just want to skip anything where `remaining_text != ""`.

Reply:
GitHub does shorten links of these forms, but their shorten text is something different.
Following is the comparison between how we do it right now and how GitHub does the same:
| Links | GitHub Version | Zulip Version |
| --- | --- | --- |
| `https://github.com/zulip/zulip/issues/11895` | ![zulip_zulip#11895](https://user-images.githubusercontent.com/67277428/111310662-5af86c80-8683-11eb-9861-ed056678bf04.png) | ![Screenshot from 2021-03-16 17-40-36](https://user-images.githubusercontent.com/67277428/111310815-824f3980-8683-11eb-8ccc-0f94f550bfdb.png) |
| `https://github.com/zulip/zulip/pull/16665` | ![zulip_zulip#16665](https://user-images.githubusercontent.com/67277428/111311003-b62a5f00-8683-11eb-8373-5efff3bdf763.png) | ![Screenshot from 2021-03-16 17-41-10](https://user-images.githubusercontent.com/67277428/111311050-c2162100-8683-11eb-8db3-0308bf81f91b.png) |
| `https://github.com/zulip/zulip/commit/620e9cbf72-ca729534aba41693d7ab2872caa394` | ![zulip_zulip@620e9cb_diff_length](https://user-images.githubusercontent.com/67277428/111311240-f984cd80-8683-11eb-8939-e56528be9acb.png) | ![Screenshot from 2021-03-16 17-43-05](https://user-images.githubusercontent.com/67277428/111311295-0c979d80-8684-11eb-93b9-9d2918f473f3.png) |
| `https://github.com/zulip/zulip/pull/16860/commits/1-9e96bcea616e6e9b1a3c10f25930ac1f75d4f92` | ![zulip_zulip@19e96bc](https://user-images.githubusercontent.com/67277428/111311456-436db380-8684-11eb-9ef6-0612731bec0b.png) | ![Screenshot from 2021-03-16 17-44-57](https://user-images.githubusercontent.com/67277428/111311496-4cf71b80-8684-11eb-8fe2-05b70b375907.png) |
| `https://github.com/zulip/zulip/pull/16665#issue-513297835` | ![zulip_zulip#16665 (comment) Initial comment](https://user-images.githubusercontent.com/67277428/111312131-01913d00-8685-11eb-858d-4415c5a34c4c.png) | ![Screenshot from 2021-03-16 17-46-42](https://user-images.githubusercontent.com/67277428/111312193-0f46c280-8685-11eb-9169-86bff6408ef4.png) |
| `https://github.com/zulip/zulip/pull/16665#issuecomment-719814618` | ![zulip_zulip#16665 (comment)](https://user-images.githubusercontent.com/67277428/111312340-2eddeb00-8685-11eb-9a72-c1022ce6db04.png) | ![Screenshot from 2021-03-16 17-47-32](https://user-images.githubusercontent.com/67277428/111312386-3bfada00-8685-11eb-8569-4e0158510df3.png) |
| `https://github.com/zulip/zulip/issues/16579#issue-725908927` | ![zulip_zulip#16579 (comment)](https://user-images.githubusercontent.com/67277428/111312585-72d0f000-8685-11eb-9bec-8ee1291a93a3.png) | ![Screenshot from 2021-03-16 17-48-12](https://user-images.githubusercontent.com/67277428/111312630-80867580-8685-11eb-9da5-b83ad9cd828f.png) |
| `https://github.com/zulip/zulip/issues/16482#issuecomment-726354516` | ![zulip_zulip#16482 (comment)](https://user-images.githubusercontent.com/67277428/111312786-a6ac1580-8685-11eb-8c3d-ebb8c4c341f6.png) | ![Screenshot from 2021-03-16 17-48-40](https://user-images.githubusercontent.com/67277428/111312881-bdeb0300-8685-11eb-9a69-d19872be4937.png) |
| `https://github.com/zulip/zulip/pull/16860/files` | ![zulip_zulip_pull_16860_files](https://user-images.githubusercontent.com/67277428/111312979-d9eea480-8685-11eb-8f5b-4ba9699a98e9.png) | ![Screenshot from 2021-03-16 17-49-19](https://user-images.githubusercontent.com/67277428/111313014-e07d1c00-8685-11eb-8107-b472597e631b.png) |

Reply:
While GitLab slightly differs from GitHub, the following is how:
| Links | GitLab Version |
| --- | --- |
| `https://gitlab.com/inkscape/inkscape/-/issues/1306` | ![Screenshot from 2021-03-16 18-43-41](https://user-images.githubusercontent.com/67277428/111314593-8ed59100-8687-11eb-8418-8e9e7de3ad59.png) |
| `https://gitlab.com/inkscape/inkscape/-/merge_requests/2688` | ![Screenshot from 2021-03-16 18-45-17](https://user-images.githubusercontent.com/67277428/111314835-c8a69780-8687-11eb-9cee-d98be19c3ae3.png) |
| `https://gitlab.com/inkscape/inkscape/-/issues/1306#note_325493115` | ![Screenshot from 2021-03-16 18-46-12](https://user-images.githubusercontent.com/67277428/111314968-ee33a100-8687-11eb-92a8-1c06f3fc6a52.png) |
| `https://gitlab.com/inkscape/inkscape/-/merge_requests/2688#note_486823380` | ![Screenshot from 2021-03-16 18-47-29](https://user-images.githubusercontent.com/67277428/111315137-17543180-8688-11eb-8913-a336efc6c790.png) |
| `https://gitlab.com/inkscape/inkscape/-/commit/9ce253cb-a0a23ef8f910bb7ab00584ccc96f9db8` | ![Screenshot from 2021-03-16 18-49-34](https://user-images.githubusercontent.com/67277428/111315485-6dc17000-8688-11eb-8ee2-8bbd4d2342ad.png) |
| `https://gitlab.com/inkscape/inkscape/-/merge_requests/2688/commits` | ![Screenshot from 2021-03-16 18-53-02](https://user-images.githubusercontent.com/67277428/111315975-df99b980-8688-11eb-9cad-d61265de1bdf.png) |
| `https://gitlab.com/inkscape/inkscape/-/merge_requests/2688/diffs?commit_id=28443c2e334c8ac4f2275e897ea116e97d595f68` | ![Screenshot from 2021-03-16 18-57-41](https://user-images.githubusercontent.com/67277428/111316771-8c743680-8689-11eb-9651-0c1df2f2f047.png) |

`GITHUB_COMMIT_ID_PREFIX_LENGTH = 7`
`GITLAB_COMMIT_ID_PREFIX_LENGTH = 8`

@timabbott, should I now try to write the code to follow GitHub and GitLab patterns exactly the same ?

Reply:
I think we should only do shortening for the basic issue, pr, and commit ones (i.e. the first 3 in your GitHub list) for the first version of this.  It'd be good to have the test suite cover the full range of examples from your comment, though, even if we're doing nothing special with these, since it'll be helpful for any additional examples.

But I basically want to be minimally aggressive on this.  Fundamentally, Zulip isn't a GitHub/GitLab-specific tool, and I think starting with a minimalist scope for what rewriting we do (and then seeing what feedback we get over the next month or two) is the right path forward.

Simple and easy to predict is best when we're not sure where we're going to ultimately land on this feature.